pilot/pkg/bootstrap/server.go

```diff
@@ -650,29 +625,67 @@ func (s *Server) createK8sServiceControllers(serviceControllers *aggregate.Contr
-func (s *Server) initMultiClusterController(args *PilotArgs) error {
+// initMultiClusterSecretController initializes multi cluster secret controller
+// to watch for remote cluster additions or deletions.  The secret controller
+// will call back to functions defined here to add handlers to watch remote clusters
+// Currently implemented only for kubernetes registries.
+func (s *Server) initMultiClusterSecretController(args *PilotArgs) (err error) {
 	if hasKubeRegistry(args) {
-		s.addStartFunc(func(stop <-chan struct{}) error {
-			secretController := clusterregistry.NewController(s.kubeClient,
-				args.Config.ClusterRegistriesNamespace,
-				s.clusterStore,
-				s.ServiceController,
-				s.EnvoyXdsServer,
-				args.Config.ControllerOptions.ResyncPeriod,
-				args.Config.ControllerOptions.WatchedNamespace,
-				args.Config.ControllerOptions.DomainSuffix)
-
-			// Start secret controller which watches for runtime secret Object changes and adds secrets dynamically
-			go secretController.Run(stop)
-
-			return nil
-		})
+
+		err = secretcontroller.StartSecretController(s.kubeClient, s.addMemberCluster, s.deleteMemberCluster, args.Config.ClusterRegistriesNamespace)
+		if err != nil {
+			return fmt.Errorf("could not start secret controller: %v", err)
+		}
+
 	}
 	return nil
 }
 
+// addMemberCluster is passed to the secret controller as a callback to be called
+// when a remote cluster is added.  This function needs to set up all the handlers
+// to watch for resources being added, deleted or changed on remote clusters.
+func (s *Server) addMemberCluster(clientset kubernetes.Interface, clusterID string) error {
+
+	//stopCh to stop and remove controller created here when cluster removed.
+	stopCh := make(chan struct{})
+	kubectl := kube.NewController(clientset, kube.ControllerOptions{
+		WatchedNamespace: s.WatchedNamespace,
+		ResyncPeriod:     s.ResyncPeriod,
+		DomainSuffix:     s.DomainSuffix,
+	})
+	s.kubeRegistry[clusterID] = kubectl
+	s.ServiceController.AddRegistry(
+		aggregate.Registry{
+			Name:             serviceregistry.KubernetesRegistry,
+			ClusterID:        clusterID,
+			ServiceDiscovery: kubectl,
+			ServiceAccounts:  kubectl,
+			Controller:       kubectl,
+		})
+
+	kubectl.StopChan = stopCh
+
+	_ = kubectl.AppendServiceHandler(func(*model.Service, model.Event) { s.EnvoyXdsServer.ClearCacheFunc()() })
+	_ = kubectl.AppendInstanceHandler(func(*model.ServiceInstance, model.Event) { s.EnvoyXdsServer.ClearCacheFunc()() })
+	go kubectl.Run(stopCh)
+
+	return nil
+}
+
+// deleteMemberCluster is passed to the secret controller as a callback to be called
+// when a remote cluster is deleted.  Also must clear the cache so remote resources
+// are removed.
+func (s *Server) deleteMemberCluster(clusterID string) error {
+
+	s.ServiceController.DeleteRegistry(clusterID)
+	kubectl := s.kubeRegistry[clusterID]
+	close(kubectl.StopChan)
+	delete(s.kubeRegistry, clusterID)
+	s.EnvoyXdsServer.ClearCacheFunc()()
+
+	return nil
+}
```

Comment:
I think these functions should be moved to a separate package/controller. The server should not be involved in setting up all the multicluster stuff. Moving them to a separate package/with its own go func or whatever, makes it easy. You simply initialize the multicluster thread and let it do its thing. The server will then be a shell that initiates various subsystems (cf, multicluster, k8s, etc..)